src/NATS.Client.Core/Telemetry.cs

```diff
@@ -0,0 +1,206 @@
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+using System.Threading.Channels;
+using Microsoft.Extensions.Primitives;
+using NATS.Client.Core.Commands;
+using OpenTelemetry.Trace;
+
+namespace NATS.Client.Core;
+
+// === Spec ====
+// - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md
+public static class Telemetry
+{
+    public const string ActivitySourceName = "NATS.Client";
+
+    public static ActivitySource NatsActivities { get; } = new(ActivitySourceName);
+
+    public static TracerProviderBuilder AddNatsInstrumentation(this TracerProviderBuilder builder)
+    {
+        builder.AddSource(ActivitySourceName);
+        return builder;
+    }
+
+    internal class Constants
```

Comment:
Need to agree on constant values. These probably should not change after release, b/c they could break monitoring rules & alerts users have set up.